internal/compiler-bridge/src/main/scala/xsbt/ExtractUsedNames.scala

```diff
@@ -78,63 +76,71 @@ class ExtractUsedNames[GlobalType <: CallbackGlobal](val global: GlobalType) ext
     }
   }
 
+  private val DefaultScopes = EnumSet.of(UseScope.Default)
+  private val PatmatScopes = EnumSet.of(UseScope.PatMatTarget)
+
   def extractAndReport(unit: CompilationUnit): Unit = {
     val tree = unit.body
     val traverser = new ExtractUsedNamesTraverser
     traverser.traverse(tree)
+
     val namesUsedAtTopLevel = traverser.namesUsedAtTopLevel
+    val defaultNamesTopLevel = namesUsedAtTopLevel.defaultNames
+    val scopedNamesTopLevel = namesUsedAtTopLevel.scopedNames
 
-    if (!namesUsedAtTopLevel.defaultNames.isEmpty || !namesUsedAtTopLevel.scopedNames.isEmpty) {
+    // Handle names used at top level that cannot be related to an owner
+    if (!defaultNamesTopLevel.isEmpty || !scopedNamesTopLevel.isEmpty) {
       val responsible = firstClassOrModuleDef(tree)
       responsible match {
         case Some(classOrModuleDef) =>
           val sym = classOrModuleDef.symbol
-          val firstClassSymbol = if (sym.isModule) sym.moduleClass else sym
+          val firstClassSymbol = enclOrModuleClass(sym)
           val firstClassName = className(firstClassSymbol)
           val namesInFirstClass = traverser.usedNamesFromClass(firstClassName)
-
-          namesInFirstClass.defaultNames.addAll(namesUsedAtTopLevel.defaultNames)
-          namesUsedAtTopLevel.scopedNames.foreach {
-            (topLevelName, topLevelScopes) =>
-              namesInFirstClass.scopedNames.get(topLevelName) match {
-                case null =>
-                  namesInFirstClass.scopedNames.put(topLevelName, topLevelScopes)
-                  ()
-                case scopes =>
-                  scopes.addAll(topLevelScopes)
-                  ()
-              }
+          val scopedNamesInFirstClass = namesInFirstClass.scopedNames
+
+          namesInFirstClass.defaultNames.addAll(defaultNamesTopLevel)
+          scopedNamesTopLevel.foreach { (topLevelName, topLevelScopes) =>
+            val existingScopes = scopedNamesInFirstClass.get(topLevelName)
+            if (existingScopes == null)
+              scopedNamesInFirstClass.put(topLevelName, topLevelScopes)
+            else existingScopes.addAll(topLevelScopes)
```

Comment:
I prefer to push changes and then discuss them, than discussing them and forcing you to spend time on it. It's just more efficient :wink:.